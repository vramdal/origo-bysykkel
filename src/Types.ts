export interface AutoDiscoveryFeedType {
  last_updated: Seconds1970;
  ttl: Seconds;
  data: {
    nb: {
      feeds: [
        {
          url: string;
          name: 'system_information';
        },
        {
          url: string;
          name: 'station_information';
        },
        {
          url: string;
          name: 'station_status';
        },
      ];
    };
  };
}

export type Seconds1970 = number;
export type Seconds = number;

export interface SystemInformationFeedType {
  last_updated: Seconds1970;
  ttl: Seconds;
  data: {
    system_id: string;
    language: string;
    name: string;
    operator: string;
    timezone: string;
    phone_number: string;
    email: string;
  };
}

export interface StationInfoType {
  station_id: string;
  name: string;
  address: string;
  lat: number;
  lon: number;
  capacity: number;
}

export interface StationInformationFeedType {
  last_updated: Seconds1970;
  ttl: Seconds;
  data: {
    stations: [StationInfoType];
  };
}

export interface StationStatusType {
  is_installed: number;
  is_renting: number;
  num_bikes_available: number;
  num_docks_available: number;
  last_reported: Seconds1970;
  is_returning: number;
  station_id: string;
}

export interface StationStatusFeedType {
  last_updated: Seconds1970;
  ttl: Seconds;
  data: {
    stations: [StationStatusType];
  };
}

export enum FetchState {
  INITIALIZING = 'INITIALIZING',
  SUCCESS = 'SUCCESS',
  FAILURE = 'FAILURE',
}
