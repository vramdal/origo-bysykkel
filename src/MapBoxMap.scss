.mapContainer {
  display: flex;
  flex: 1 0 auto;
  overflow: hidden;

  .station-marker {
    background-color: ButtonFace;
    padding: 1em;
    border-radius: 5px;
    color: black;
    max-width: 5em;
    transition: max-width .1s, background-color .2s;

    &.zoom-WORLD {
      background-color: transparent;
      max-width: 25px;
    }

    div {
      white-space: nowrap;
      text-overflow: ellipsis;
      overflow: hidden;
    }

    .oslobysykkel-logo {
      width: 25px;
    }

    &:not(.has-bikes-available) {
      @extend .lock-no-bike-available
    }

    &:not(.has-locks-available) {
      @extend .bike-no-lock-available
    }

    &:not(.has-locks-available):not(.has-bikes-available) {
     @extend .no-bike-no-lock-available
    }

    h2 {
      font-size: 10pt;
      white-space: pre;
      overflow: hidden;
      text-overflow: ellipsis;
    }

    &.zoom-DETAIL {
      max-width: 10em;

      h2 {
        white-space: normal;
        overflow: visible;

      }
    }

    .bikes, .locks {
      font-size: 14pt;
    }
  }
}

.legend {
  flex: 0 0 5px;
  display: flex;
  font-size: 10pt;
  padding: 0.3em 1em;
  justify-content: space-evenly;

  > div {
    display: flex;
    flex: 1 1 auto;
    justify-content: center;

    > img {
      margin-right: 5px;
    }
  }

  img {
    height: 15px;
  }
}

.bike-no-lock-available {
  filter: hue-rotate(180deg);
}

.lock-no-bike-available {
  filter: hue-rotate(90deg);
}

.no-bike-no-lock-available {
  filter: grayscale(100%);
}