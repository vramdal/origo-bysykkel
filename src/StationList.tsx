import { Cell, ColumnHeader, Row, RowGroup, Table } from './Layout';
import classnames from 'classnames';
import { FetchState } from './Types';
import React from 'react';

export type DisplayStationStatus = {
  stationId: string;
  stationName: string;
  numBikes?: number;
  numLocks?: number;
  isReturning: boolean;
  isRenting: boolean;
  lat: number;
  lon: number;
};

export type DisplayStationStatusList = Array<DisplayStationStatus>;
const StationRow = ({ station }: { station: DisplayStationStatus }): JSX.Element => (
  <Row>
    <Cell>{station.stationName}</Cell>
    <Cell className={classnames({ disabled: !station.isReturning, exhausted: station.numLocks === 0 })}>
      {station.numLocks}
    </Cell>
    <Cell className={classnames({ disabled: !station.isRenting, exhausted: station.numBikes === 0 })}>
      {station.numBikes}
    </Cell>
  </Row>
);

interface StationListProps {
  stations: DisplayStationStatusList;
  fetchState: FetchState;
}

export const StationList = (props: StationListProps): JSX.Element => (
  <Table className={classnames('station-table', { disabled: props.fetchState === FetchState.FAILURE })}>
    <Row className={'station-table__header'}>
      <ColumnHeader>Stasjon</ColumnHeader>
      <ColumnHeader>Antall l√•ser</ColumnHeader>
      <ColumnHeader>Antall sykler</ColumnHeader>
    </Row>
    <RowGroup className={'station-table__body'}>
      {props.stations?.map((station: DisplayStationStatus) => (
        <StationRow key={station.stationId} station={station} />
      ))}
    </RowGroup>
  </Table>
);
