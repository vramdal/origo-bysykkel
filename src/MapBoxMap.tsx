import React, { RefObject } from 'react';
import mapboxgl from 'mapbox-gl';
import { FetchState } from './Types';
import { DisplayStationStatus, DisplayStationStatusList } from './StationList';
import './MapBoxMap.scss';
import classNames from 'classnames';
import logo from './oslobysykkel-logo.png';

interface ComponentState {
  center: {
    lng: number;
    lat: number;
  };
  zoom: number;
  geolocation?: {
    lng: number;
    lat: number;
  };
}

export interface MapBoxMapProps {
  fetchState: FetchState;
  stations: DisplayStationStatusList;
}

enum ZoomCategory {
  WORLD = 'WORLD',
  MEDIUM = 'MEDIUM',
  DETAIL = 'DETAIL',
}

const categorizeZoom = (zoom: number): ZoomCategory => {
  if (zoom > 18) {
    return ZoomCategory.DETAIL;
  } else if (zoom > 14) {
    return ZoomCategory.MEDIUM;
  } else {
    return ZoomCategory.WORLD;
  }
};

export class MapBoxMap extends React.Component<MapBoxMapProps, ComponentState> {
  static defaultProps = {
    stations: [],
  };

  constructor(props: MapBoxMapProps) {
    super(props);
    mapboxgl.accessToken = 'pk.eyJ1IjoidnJhbWRhbCIsImEiOiJjaXVmNW5qMncwMDE1MnlwcmN1MnJkMzQ4In0.4_QLMQFAPRz_vfwn5m76vQ';
    this.state = {
      center: {
        lng: 10.7303766,
        lat: 59.9108469,
      },
      zoom: 15,
    };
    for (const station of props.stations) {
      const id = station.stationId;
      const ref: React.RefObject<HTMLDivElement> = React.createRef();
      this.markerRefs.set(id, ref);
    }
  }

  private mapContainerRef: RefObject<HTMLDivElement> = React.createRef();
  private map?: mapboxgl.Map = undefined;
  private markers: Map<string, mapboxgl.Marker> = new Map<string, mapboxgl.Marker>();
  private markerRefs: Map<string, React.RefObject<HTMLDivElement>> = new Map();

  componentDidMount(): void {
    if (this.mapContainerRef.current) {
      this.map = new mapboxgl.Map({
        container: this.mapContainerRef.current,
        style: 'mapbox://styles/mapbox/streets-v11',
        center: [this.state.center.lng, this.state.center.lat],
        zoom: this.state.zoom,
      });
      this.map.addControl(new mapboxgl.NavigationControl());
      const geolocate = new mapboxgl.GeolocateControl({
        positionOptions: {
          enableHighAccuracy: true,
        },
        showAccuracyCircle: false,
        fitBoundsOptions: {
          maxZoom: 18,
        },
        trackUserLocation: true,
      });
      this.map.addControl(geolocate);
      this.map.on('zoomend', () => {
        const zoom = this.map?.getZoom() || 15;
        this.setState({ zoom: zoom });
      });
      this.map.on('load', function () {
        geolocate.trigger();
      });
      this.createMarkers(this.props.stations, this.map);
    }
  }

  createMarkers(stations: DisplayStationStatusList, map: mapboxgl.Map): void {
    for (const station of stations) {
      const id = station.stationId;
      const ref = this.markerRefs.get(id);
      if (ref) {
        const marker = new mapboxgl.Marker({ element: ref.current || undefined });
        marker.setLngLat([station.lon, station.lat]);
        this.markers.set(id, marker);
        marker.addTo(map);
      }
    }
  }

  zoomToStation(station: DisplayStationStatus): void {
    this.map?.easeTo({ center: { lat: station.lat, lng: station.lon }, animate: true, zoom: 19 });
  }

  render(): JSX.Element {
    const zoomCategory: string = categorizeZoom(this.state.zoom);
    const stationElements = this.props.stations.map((station) => {
      return (
        <div key={station.stationId} ref={this.markerRefs.get(station.stationId)}>
          <div
            className={classNames('station-marker', `zoom-${zoomCategory}`, {
              'has-bikes-available': station.numBikes && station.numBikes > 0,
              'has-locks-available': station.numLocks && station.numLocks > 0,
            })}
            onClick={(): void => this.zoomToStation(station)}
            role={'tooltip'}
          >
            <img src={logo} alt={'Oslo Bysykkel-logo'} className={'oslobysykkel-logo'} />
            {zoomCategory === ZoomCategory.DETAIL && (
              <>
                <h2>{station.stationName}</h2>
                {station.stationName !== station.address && <p>{station.address}</p>}
              </>
            )}
            {(zoomCategory === ZoomCategory.DETAIL || zoomCategory === ZoomCategory.MEDIUM) && (
              <>
                <div className={'bikes'}>
                  <span role={'img'} aria-label={'sykkel'}>
                    ðŸš²
                  </span>{' '}
                  {station.numBikes}
                </div>
                <div className={'locks'}>
                  <span role={'img'} aria-label={'sykkel'}>
                    ðŸ”’
                  </span>{' '}
                  {station.numLocks}
                </div>
              </>
            )}
          </div>
        </div>
      );
    });
    return (
      <div ref={this.mapContainerRef} className="mapContainer" data-testid={'mapContainer'} role="application">
        {stationElements}
      </div>
    );
  }
}
