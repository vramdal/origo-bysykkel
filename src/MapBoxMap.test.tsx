import { _testing, MapBoxMap } from './MapBoxMap';

import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import { DisplayStationStatus } from './StationList';
import { when } from 'jest-when';
import mapboxgl from 'mapbox-gl';
import { FetchState } from './Types';

const originalMapboxGlModule = jest.requireActual('mapbox-gl');
const originalMarker = originalMapboxGlModule.Marker;
jest.mock('mapbox-gl', () => {
  const originalModule = jest.requireActual('mapbox-gl');
  return {
    ...originalModule,
    Map: jest.fn(),
    Marker: jest.fn(),
  };
});
const mockMap = mapboxgl.Map as jest.Mock;
const mockMarker = mapboxgl.Marker as jest.Mock;

const { StationMapMarker, ZoomCategory } = _testing;

const station1: DisplayStationStatus = {
  lon: 1,
  lat: 2,
  stationId: 'station-1',
  numLocks: 5,
  numBikes: 8,
  stationName: 'Station 1',
  address: 'Street 1',
  isRenting: true,
  isReturning: true,
};

const station2: DisplayStationStatus = {
  lon: 2,
  lat: 3,
  stationId: 'station-2',
  numLocks: 11,
  numBikes: 7,
  stationName: 'Station 2',
  address: 'Street 2',
  isRenting: true,
  isReturning: true,
};

describe('MapBox integration', () => {
  it('should ', async () => {
    const addedMarkers = [];

    const addMarker = jest.fn();
    addMarker.mockImplementation(() => {
      addedMarkers.push(addMarker.mock.instances[0]);
    });
    const mockMapImplementation = {
      addControl: jest.fn(),
      on: jest.fn(),
      easeTo: jest.fn(),
      addMarker: addMarker,
    };

    when(mockMap).calledWith(expect.anything()).mockReturnValue(mockMapImplementation);
    when(mockMarker)
      .calledWith(expect.anything())
      .mockImplementation((...args) => {
        const marker = new originalMarker(...args);
        marker.addTo = mockMapImplementation.addMarker;
        return marker;
      });

    const { container } = render(<MapBoxMap fetchState={FetchState.SUCCESS} stations={[station1, station2]} />);

    expect(mockMapImplementation.on).toHaveBeenCalledWith('load', expect.any(Function));
    expect(mockMapImplementation.on).toHaveBeenCalledWith('zoomend', expect.any(Function));
    expect(mockMapImplementation.addControl).toHaveBeenCalledWith(expect.any(mapboxgl.NavigationControl));
    expect(mockMapImplementation.addControl).toHaveBeenCalledWith(expect.any(mapboxgl.GeolocateControl));
    expect(mockMarker).toHaveBeenCalledTimes(2);
    expect(mockMapImplementation.addMarker).toHaveBeenCalledTimes(2);
    expect(container).toMatchSnapshot();
  });
});

describe('Stations on the map', () => {
  const clickHandler = jest.fn();

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should set a class names based on zoom level and availability', () => {
    const { getByRole } = render(
      <StationMapMarker station={station1} zoomCategory={ZoomCategory.DETAIL} markerClickHandler={clickHandler} />,
    );

    const element = getByRole('tooltip');
    expect(element).toHaveClass(`zoom-DETAIL has-bikes-available has-locks-available`);
  });

  it('should call clickHandler when clicked', () => {
    const { getByRole } = render(
      <StationMapMarker station={station1} zoomCategory={ZoomCategory.DETAIL} markerClickHandler={clickHandler} />,
    );
    const element = getByRole('tooltip');

    fireEvent.click(element);

    expect(clickHandler).toHaveBeenCalled();
  });

  it('should not set has-bikes-available', () => {
    const { getByRole } = render(
      <StationMapMarker
        station={{ ...station1, numBikes: 0 }}
        zoomCategory={ZoomCategory.DETAIL}
        markerClickHandler={clickHandler}
      />,
    );

    const element = getByRole('tooltip');

    expect(element).toHaveClass(`zoom-DETAIL has-locks-available`);
    expect(element).not.toHaveClass(`has-bikes-available`);
  });

  it('should not set has-locks-available', () => {
    const { getByRole } = render(
      <StationMapMarker
        station={{ ...station1, numLocks: 0 }}
        zoomCategory={ZoomCategory.DETAIL}
        markerClickHandler={clickHandler}
      />,
    );

    const element = getByRole('tooltip');

    expect(element).toHaveClass(`zoom-DETAIL has-bikes-available`);
    expect(element).not.toHaveClass(`has-locks-available`);
  });

  it('should render name and address when zoomed to detail level', () => {
    const { getByRole } = render(
      <StationMapMarker
        station={{ ...station1, numLocks: 0 }}
        zoomCategory={ZoomCategory.DETAIL}
        markerClickHandler={clickHandler}
      />,
    );

    const element = getByRole('tooltip');
    const title = element.querySelector('h2');
    const address = element.querySelector('address');
    expect(title).toHaveTextContent('Station 1');
    expect(address).toHaveTextContent('Street 1');
  });
});
