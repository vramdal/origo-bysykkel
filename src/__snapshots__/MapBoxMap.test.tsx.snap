// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MapBox integration should  1`] = `
<div>
  <div
    class="mapContainer"
    data-testid="mapContainer"
    role="application"
  >
    <div
      class="mapboxgl-marker mapboxgl-marker-anchor-center"
    >
      <div
        class="station-marker zoom-MEDIUM has-bikes-available has-locks-available"
        role="tooltip"
      >
        <img
          alt="Oslo Bysykkel-logo"
          class="oslobysykkel-logo"
          src="oslobysykkel-logo.png"
        />
        <div
          class="bikes"
        >
          <span
            aria-label="sykkel"
            role="img"
          >
            ðŸš²
          </span>
           
          8
        </div>
        <div
          class="locks"
        >
          <span
            aria-label="sykkel"
            role="img"
          >
            ðŸ”’
          </span>
           
          5
        </div>
      </div>
    </div>
    <div
      class="mapboxgl-marker mapboxgl-marker-anchor-center"
    >
      <div
        class="station-marker zoom-MEDIUM has-bikes-available has-locks-available"
        role="tooltip"
      >
        <img
          alt="Oslo Bysykkel-logo"
          class="oslobysykkel-logo"
          src="oslobysykkel-logo.png"
        />
        <div
          class="bikes"
        >
          <span
            aria-label="sykkel"
            role="img"
          >
            ðŸš²
          </span>
           
          7
        </div>
        <div
          class="locks"
        >
          <span
            aria-label="sykkel"
            role="img"
          >
            ðŸ”’
          </span>
           
          11
        </div>
      </div>
    </div>
  </div>
  <div
    class="legend"
    role="list"
  >
    <div
      role="listitem"
    >
      <img
        alt="Oslo Bysykkel-logo"
        class="oslobysykkel-logo bike-and-lock-available"
        src="oslobysykkel-logo.png"
      />
      <span>
        Sykkel og lÃ¥s ledig
      </span>
    </div>
    <div>
      <img
        alt="Oslo Bysykkel-logo"
        class="oslobysykkel-logo bike-no-lock-available"
        src="oslobysykkel-logo.png"
      />
      <span>
        Ingen lÃ¥s ledig
      </span>
    </div>
    <div>
      <img
        alt="Oslo Bysykkel-logo"
        class="oslobysykkel-logo lock-no-bike-available"
        src="oslobysykkel-logo.png"
      />
      <span>
        Ingen sykkel ledig
      </span>
    </div>
  </div>
</div>
`;
